### PR TITLE
MR-3373: Code coverage explainer

### PR DESCRIPTION
**Pull request by rswerve**
## Summary

The way we generate our combined code coverage report is not obvious, so I wrote up this explainer.

## QA guidance

Is it clear?  Do you now understand our process?

### CODE REVIEW

**Review by mojotalantikite — APPROVED** [review]

**Review by macrael — APPROVED** [review]

File: docs/technical-design/code_coverage.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+The `nyc` commands to combine and publish reports in CI are the same as those we use locally, but there are some wrinkles to be aware of.
+
+In CI, we run our Cypress tests in parallel in order to save time. Under the `Cypress` key in deploy.yml, `matrix` has a subkey set to an array. The length of the array determines the number of parallel processes that will be spawned. Each process will generate its own coverage report. Those coverage reports themselves need to be merged into one before they can be merged with the unit test reports.
+
+Under the `Create combined test coverage report` key in deploy.yml, after some mkdir setup, you can see that we copy all the generated Cypress coverage files into one directory, and then run `npx nyc merge` on that directory. (`npx` is used to run commands from npm packages without permanently installing them.)
+
+Next we copy the unit test reports into the same directory, and run `npx nyc merge` again. Then, as in our package.json, we generate the reports.
+
+These reports are then uploaded as artifacts (under the `Upload combined test coverage` key), which can be downloaded from the github action summary for the deployment.
```
Comment (haworku) [inline docs/technical-design/code_coverage.md:74]:
```suggestion
These reports are then uploaded as [workflow artifacts](https://docs.github.com/en/actions/managing-workflow-runs/downloading-workflow-artifacts) under `Upload combined test coverage`. This zip file can be downloaded from the github action summary for the `deploy` workflow (scroll to bottom of page).
```

File: docs/technical-design/code_coverage.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# Code coverage reports
+
+## Overview
+
+Code coverage is a measure of the percentage of code that is exercised during testing. Detailed coverage reports can identify which specific lines of code are being tested. In order for coverage to be measured, the code must first be instrumented. Tools that instrument code effectively wrap it in functions that increment a counter when that code is run. Code coverage generators take those counts and turn them into reports about covered and uncovered lines of code.
+
```
Comment (haworku) [inline docs/technical-design/code_coverage.md:6]: Is this also what feed the code coverage metrics in Code Climate? If so, I feel like that is worth mentioning here.

File: docs/technical-design/code_coverage.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+Under the `Create combined test coverage report` key in deploy.yml, after some mkdir setup, you can see that we copy all the generated Cypress coverage files into one directory, and then run `npx nyc merge` on that directory. (`npx` is used to run commands from npm packages without permanently installing them.)
+
+Next we copy the unit test reports into the same directory, and run `npx nyc merge` again. Then, as in our package.json, we generate the reports.
+
+These reports are then uploaded as artifacts (under the `Upload combined test coverage` key), which can be downloaded from the github action summary for the deployment.
+
+### notes
+
+[^1]: This is trivial to do in Jest by passing a `--coverage` flag, and more involved in Cypress, as implemented in parts of [this pull request](https://github.com/Enterprise-CMCS/managed-care-review/pull/1119/files).
```
Comment (haworku) [inline docs/technical-design/code_coverage.md:78]: This is good - I feel like it could be easy to miss this part ⭐

**Review by haworku — APPROVED** [review]

File: docs/technical-design/code_coverage.md (hunk 2 above)
Comment (rswerve) [reply docs/technical-design/code_coverage.md:6]: Good call.  Added that bit later in the doc.

**Pull request merged**